### PR TITLE
Add typespecs to Sequence, MandelbrotSet and Complex

### PR DESCRIPTION
**Pull request by munckymagik**
# TL;DR

Add typespecs and run dialyzer in order to:
- Find bugs your tests might not
- Give a little extra information to developers reading your code for understanding 
# What?

My first time using typespecs in Elixir.

There are Mix plugins that make it easier to do this, but I opted for the manual approach this time with the hope of learning more.
# Why?

Elixir itself doesn't do anything with the information added by the typespecs. However beyond just being extra documentation, they make it possible to do **static analysis of your code**.

You can use the [Dialyzer](http://www.erlang.org/doc/apps/dialyzer/dialyzer_chapter.html) tool (bundled with Erlang) to actually validate your code with the specs you add.

This provides potential to find certain kinds of issues that your tests might not tease out. E.g. weeding out dead or unreachable code, unnecessary tests, etc

See the [Dialyzer docs](http://www.erlang.org/doc/man/dialyzer.html) for more info. 
# Annotating the code

I'll not go into detail here, it's well documented. I found these links useful:
- [Elixir Typespecs Guide](http://elixir-lang.org/getting-started/typespecs-and-behaviours.html) - Gets you started
- [Elixir API Docs Page on Typespecs](http://elixir-lang.org/docs/stable/elixir/typespecs.html) -This one lists the available built-in type names you can use
# Validating the code
## Step 1: create a PST file for Dialyzer

Dialyzer supports creation of a PST (Persistent Lookup Table) file to cache the results of analysing the Erlang/Elixir libraries so you don't have to wait while it does this every time you analyse your own code.

First create the file and by analysing Erlang's libraries:

```
$ dialyzer --build_plt --apps erts kernel stdlib crypto public_key
  Compiling some key modules to native code... done in 0m33.06s
  Creating PLT /Users/danm/.dialyzer_plt ...
Unknown functions:
  asn1rt_nif:decode_ber_tlv/1
  asn1rt_nif:encode_ber_tlv/1
  compile:file/2
  compile:forms/2
  compile:noenv_forms/2
  compile:output_generated/1
Unknown types:
  compile:option/0
 done in 1m6.38s
done (passed successfully)
```

On my machine this took under a minute to run and outputs to `~/.dialyzer_plt` by default. At this point the file is 780Kb.

Next I added the Elixir libraries:

```
$ dialyzer --add_to_plt --plt ~/.dialyzer_plt -r /usr/local/Cellar/elixir/1.2.0/lib
```

At this point the PLT file grows to 1.0MB.
## Step 2: analyse the code

First, compile the project so there is some code to analyse:

```
$ mix compile
```

Then run the analysis by executing dialyzer passing the path to folder that contains your code's compiled binary beam files:

```
$ dialyzer _build/dev/lib/mandelbrot/ebin/
  Checking whether the PLT /Users/danm/.dialyzer_plt is up-to-date... yes
  Proceeding with analysis...
sequence.ex:11: Invalid type specification for function 'Elixir.Sequence':finite_lazy/3. The success typing is (pos_integer(),number(),number()) -> #{}
 done in 0m1.05s
done (warnings were emitted)
```

---

**UPDATE** actually considering we want Dialyzer to look at our test code too, I'm assuming we should run it like:

```
$ mix test && dialyzer _build/test/lib/mandelbrot/ebin/
```

... although I've yet to find a way to break the tests from Dialyzer's perspective so I'm not sure it's actually looking at them yet.

---
## Step 3: interpret and apply fixes for warnings

At the end of the previous step in my example, this warning was output:

> sequence.ex:11: Invalid type specification for function 'Elixir.Sequence':finite_lazy/3. The success typing is (pos_integer(),number(),number()) -> #{}

The signature for the offending function looks like:

``` elixir
@spec finite_lazy(pos_integer, number, number) :: Stream.t
```

We don't strictly need the parantheses `()` after each type name, so the issue here seems to be about the return type. I've said I expect it to return a `Stream.t` because the last call in the function is to `Stream.map`. But I must have guessed wrongly. It's saying if I change the type to be a `map` it will work.

At this point I should say, remember: **any fixes suggested by the output will be in Erlang syntax!!!!**. So I cannot simply change it to:

``` elixir
# WRONG, Erlang literal syntax for a map
@spec finite_lazy(pos_integer, number, number) :: #{}

# CORRECT, Elixir literal syntax
@spec finite_lazy(pos_integer, number, number) :: %{}
```

So I replace `Stream.t` with `%{}` and recompile (**IMPORTANT** recompile. So many times I have made changes and just re-ran dialyzer only to be frustrated there is no change. It runs on your beam files so remember to `mix compile` after any change). Success, this does indeed make the warning go away.

However, saying it returns a map is a bit of a cop out. The map is the underlying type returned, so it is strictly correct because it's the most generic thing that can be said about the return type. But this communicates very little to developers reading for understanding (and possibly not enough to help dialyzer strongly enforce usage of the return type, although I'm not experienced enough with it yet to know).

I had assumed that because there was a [`Stream.t` defined](https://github.com/elixir-lang/elixir/blob/v1.2.0/lib/elixir/lib/stream.ex#L99), that this is the type that best communicated what was returned from `Stream.map`. But it seems I was wrong. Looking at the [docs for `Stream.map`](http://elixir-lang.org/docs/stable/elixir/Stream.html#map/2), it says it returns an `Enumerable.t`. So I update my code to the following:

``` elixir
@spec finite_lazy(pos_integer, number, number) :: Enumerable.t
```

A quick recompile and rerun of dialyzer confirms that this is correct.

```
$ mix compile && dialyzer _build/dev/lib/mandelbrot/ebin/
Compiled lib/sequence.ex
  Checking whether the PLT /Users/danm/.dialyzer_plt is up-to-date... yes
  Proceeding with analysis... done in 0m1.02s
done (passed successfully)
```

### CODE REVIEW

**Comment by paulanthonywilson** [comment]
I think this would make a valuable blog post. Could do with
- a bit of an introduction on what type specs are and why it's a good idea to use them, then link out to the typespec docs
- same for dialyzer

That's it really.

**Comment by paulanthonywilson** [comment]
Maybe a toy example, to follow along with. Or maybe not.

**Comment by munckymagik** [comment]
Great thanks for the feedback :+1:
